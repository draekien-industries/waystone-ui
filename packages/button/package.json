{
  "name": "@waystone/button",
  "version": "0.0.0",
  "main": "./src/index.ts",
  "module": "./src/index.ts",
  "types": "./src/index.ts",
  "private": true,
  "repository": {
    "type": "git",
    "url": "https://github.com/draekien-industries/waystone-ui"
  },
  "description": "The HTML Button component for @waystone/ui",
  "author": "draekien-industries",
  "sideEffects": false,
  "license": "MIT",
  "scripts": {
    "lint": "eslint --ext .tsx,.ts,.jsx,.js ./src/ --fix",
    "lint:release": "eslint --ext .tsx,.ts,.jsx,.js ./src/",
    "test": "jest --coverage --passWithNoTests",
    "test:release": "jest --coverage --passWithNoTests",
    "test:update": "jest --coverage --passWithNoTests -u",
    "test:watch": "jest --coverage --passWithNoTests --watch -u"
  },
  "devDependencies": {
    "@testing-library/jest-dom": "^6.4.2",
    "@testing-library/react": "^14.2.2",
    "@theme-ui/color": "^0.16.2",
    "@waystone/core": "workspace:*",
    "@waystone/utils": "workspace:*",
    "@waystone/icon": "workspace:*",
    "@waystone/testing-library": "workspace:*",
    "@waystone/tsconfig": "workspace:*",
    "eslint-config-waystone": "workspace:*",
    "eslint-plugin-waystone": "workspace:*",
    "react": "^18.2.0",
    "theme-ui": "^0.16.2"
  }
}
