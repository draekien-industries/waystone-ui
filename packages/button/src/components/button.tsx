import { Icon } from '@waystone/icon';
import {
  Box,
  Button as ThemeUIButton,
  type ThemeUIStyleObject,
} from 'theme-ui';
import {
  outlineCss,
  boxShadowCss,
  disabledCss,
  cssSelectors,
} from '@waystone/css-presets';
import { forwardRef, type ForwardedRef } from 'react';
import type { ButtonProps, ButtonSize } from './button.types';
import {
  getActiveBackgroundColor,
  getBackgroundColor,
  getColor,
  getHoverBackgroundColor,
  getInteractiveColor,
  paddingY,
} from './button.styles';

const renderButton = (
  {
    children,
    icon,
    loading,
    size = 'small',
    variant = 'primary',
    color,
    fullWidth,
    ...rest
  }: ButtonProps,
  ref: ForwardedRef<HTMLButtonElement>
) => {
  const disabled = loading || rest.disabled;
  const isLink = variant === 'link';
  const isGhost = variant === 'ghost';

  const interactiveColor = getInteractiveColor(variant);

  const hoverCss: ThemeUIStyleObject = {
    color: interactiveColor,
    boxShadow: boxShadowCss.focused,
    backgroundColor: getHoverBackgroundColor(variant),
  };

  const focusCss: ThemeUIStyleObject = {
    ...hoverCss,
    boxShadow: boxShadowCss.active,
  };

  const activeCss: ThemeUIStyleObject = {
    ...focusCss,
    backgroundColor: getActiveBackgroundColor(variant),
  };

  return (
    <ThemeUIButton
      ref={ref}
      sx={{
        display: 'flex',
        flexWrap: 'nowrap',
        alignItems: 'center',
        justifyContent: 'center',
        color: color || getColor(variant),
        gap: 'small',
        paddingX: isLink ? 0 : size,
        paddingY: isLink ? 0 : paddingY[size],
        width: fullWidth ? '100%' : undefined,
        cursor: 'pointer',
        transition: 'all 200ms',
        backgroundColor: rest.active
          ? activeCss.backgroundColor
          : getBackgroundColor(variant),
        ...outlineCss.base,
        ...(isLink || isGhost ? {} : boxShadowCss.base),
        [cssSelectors.hover]: {
          ...hoverCss,
        },
        [cssSelectors.focusVisible]: {
          ...outlineCss.focused,
          ...focusCss,
        },
        [cssSelectors.active]: {
          ...outlineCss.active,
          ...activeCss,
        },
        [cssSelectors.disabled]: {
          ...disabledCss,
          backgroundColor: isLink ? undefined : 'muted',
          cursor: 'not-allowed',
        },
      }}
      {...rest}
      disabled={disabled || loading}
    >
      {loading ? (
        <Box>Loading...</Box> // TODO: loading state
      ) : (
        <>
          {icon && <Icon {...icon}>{icon.name}</Icon>}
          {children}
        </>
      )}
    </ThemeUIButton>
  );
};

export const Button = forwardRef(renderButton);
