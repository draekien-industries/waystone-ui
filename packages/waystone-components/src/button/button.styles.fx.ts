import { addDarkModeSuffix, ButtonVariant, Size } from '@waystone/core';

type getBackgroundColorProps = {
  variant?: ButtonVariant;
  darkMode?: boolean;
};

export const getBackgroundColor = ({
  variant = 'primary',
}: getBackgroundColorProps) => {
  switch (variant) {
    case 'ghost':
      return 'ghost';
    case 'link':
      return 'transparent';
    case 'primary':
    case 'secondary':
    case 'accent':
    case 'highlight':
    default:
      return `${variant[0]}-400`;
  }
};

/**
 * If the variant is ghost, return muted, if it's link, return transparent, otherwise return the
 * variant with a -100 suffix.
 * @param {ButtonVariant} variant - ButtonVariant
 * @returns A function that takes a variant and returns a color.
 */
export const getHoverBackgroundColor = ({
  variant = 'primary',
  darkMode,
}: getBackgroundColorProps) => {
  switch (variant) {
    case 'ghost':
      return darkMode ? 'b-600' : 'b-200';
    case 'link':
      return 'transparent';
    case 'primary':
    case 'secondary':
    case 'accent':
    case 'highlight':
    default:
      return `${variant[0]}-300`;
  }
};

/**
 * If the variant is ghost, return muted, if it's link, return transparent, otherwise return the
 * variant with a 600 appended to it.
 * @param {ButtonVariant} variant - ButtonVariant
 * @returns A function that takes a variant and returns a color.
 */
export const getActiveBackgroundColor = ({
  variant = 'primary',
  darkMode,
}: getBackgroundColorProps) => {
  switch (variant) {
    case 'ghost':
      return darkMode ? 'b-500' : 'b-50';
    case 'link':
      return 'transparent';
    case 'primary':
    case 'secondary':
    case 'accent':
    case 'highlight':
    default:
      return `${variant[0]}-200`;
  }
};

/**
 * It returns a string based on the variant passed in
 * @param {ButtonVariant} variant - The variant of the button.
 * @returns A string
 */
export const getColor = (variant?: ButtonVariant) => {
  switch (variant) {
    case 'primary':
    case 'secondary':
    case 'accent':
    case 'highlight':
      return 'b-800';
    default:
      return 'text';
  }
};

/**
 * Gets the active color for a link button based on the dark mode
 * @param darkMode - Is the app using dark mode?
 * @returns
 */
export const getLinkVariantActiveColor = (darkMode: boolean) =>
  darkMode ? 'h-400' : 'h-700';

/**
 * Gets the hover color for a link button based on the dark mode
 * @param darkMode - Is the app using dark mode?
 * @returns
 */
export const getLinkVariantHoverColor = (darkMode: boolean) =>
  darkMode ? 'h-300' : 'h-600';

/**
 * `ComputedBoxShadowStyles` is an object with three properties: `normal`, `hover`, and `active`. Each
 * of those properties is a string.
 * @property {string} normal - The box-shadow style for the normal state.
 * @property {string} hover - The box-shadow style when the user hovers over the button.
 * @property {string} active - The box-shadow style when the button is active.
 */
type ComputedBoxShadowStyles = {
  normalBoxShadow: string;
  hoverBoxShadow: string;
  activeBoxShadow: string;
};

type getBoxShadowProps = {
  variant?: ButtonVariant;
  darkMode?: boolean;
};

export const getBoxShadow = ({
  variant = 'primary',
  darkMode,
}: getBoxShadowProps): ComputedBoxShadowStyles => {
  if (['ghost', 'link'].includes(variant)) {
    return {
      normalBoxShadow: 'none',
      hoverBoxShadow: 'none',
      activeBoxShadow: 'none',
    };
  }

  return {
    normalBoxShadow: addDarkModeSuffix('sm', darkMode),
    hoverBoxShadow: addDarkModeSuffix('md', darkMode),
    activeBoxShadow: addDarkModeSuffix('sm', darkMode),
  };
};

/**
 * `ComputedPaddingStyles` is an object with two properties, `paddingX` and `paddingY`, which are both
 * strings or numbers.
 * @property {string | number} paddingX - The padding on the left and right sides of the component.
 * @property {string | number} paddingY - The padding on the top and bottom of the component.
 */
type ComputedPaddingStyles = {
  paddingX: string | number;
  paddingY: string | number;
};

type getPaddingProps = {
  variant?: ButtonVariant;
  size?: Size;
};

export const getPadding = ({
  variant = 'primary',
  size,
}: getPaddingProps): ComputedPaddingStyles => {
  if (variant === 'link') return { paddingX: 0, paddingY: 0 };

  switch (size) {
    case 'sm': {
      return {
        paddingX: 'sm',
        paddingY: 'xs',
      };
    }
    case 'lg': {
      return {
        paddingX: 'xl',
        paddingY: 'sm',
      };
    }
    default:
      return {
        paddingX: 'lg',
        paddingY: 'sm',
      };
  }
};
