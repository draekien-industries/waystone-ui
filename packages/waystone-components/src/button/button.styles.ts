import { alpha } from '@theme-ui/color';
import { ButtonVariant, Size } from '@waystone/core';
import { ThemeUIStyleObject } from 'theme-ui';
import {
  CanActivate,
  CanBeFullWidth,
  HasSize,
  HasVariant,
  HasWidth,
} from '../common';
import {
  getActiveBackgroundColor,
  getBackgroundColor,
  getBoxShadow,
  getColor,
  getHoverBackgroundColor,
  getLinkVariantActiveColor,
  getLinkVariantHoverColor,
  getPadding,
} from './button.styles.fx';

export interface ButtonCssParams
  extends HasVariant<ButtonVariant>,
    HasSize,
    HasWidth,
    CanBeFullWidth,
    CanActivate {
  darkMode: boolean;
}

export const buttonCss = ({
  variant,
  size,
  darkMode,
  width,
  minWidth,
  maxWidth,
  fullWidth,
  active,
}: ButtonCssParams): ThemeUIStyleObject => {
  const { normalBoxShadow, hoverBoxShadow, activeBoxShadow } = getBoxShadow({
    variant,
    darkMode,
  });

  const { paddingX, paddingY } = getPadding({ variant, size });
  const color = getColor(variant);
  const outlineColor = darkMode ? 'b-600' : 'info-000';

  const isLink = variant === 'link';

  const hoverCss: ThemeUIStyleObject = {
    color: isLink ? getLinkVariantHoverColor(darkMode) : color,
    boxShadow: hoverBoxShadow,
    backgroundColor: getHoverBackgroundColor({ variant, darkMode }),
  };

  const focusCss: ThemeUIStyleObject = {
    color: isLink ? getLinkVariantActiveColor(darkMode) : color,
    boxShadow: activeBoxShadow,
    backgroundColor: getActiveBackgroundColor({ variant, darkMode }),
    outlineColor: alpha(outlineColor, 1),
    outlineWidth: 'sm',
    outlineStyle: 'solid',
  };

  const activeCss: ThemeUIStyleObject = {
    color: isLink ? getLinkVariantActiveColor(darkMode) : color,
    boxShadow: activeBoxShadow,
    backgroundColor: getActiveBackgroundColor({ variant, darkMode }),
    outlineColor: alpha(outlineColor, 0),
    outlineWidth: 'sm',
    outlineStyle: 'solid',
  };

  const baseCss: ThemeUIStyleObject = {
    display: 'flex',
    flexWrap: 'nowrap',
    alignItems: 'center',
    justifyContent: 'center',
    color,
    backgroundColor: getBackgroundColor({ variant, darkMode }),
    paddingX,
    paddingY,
    width: fullWidth ? '100%' : width,
    maxWidth: !fullWidth && maxWidth,
    minWidth: !fullWidth && minWidth,
    boxShadow: normalBoxShadow,
    margin: 0,
    cursor: 'pointer',
    transition: 'all 200ms ease',
    outlineColor: alpha(outlineColor, 0),
    outlineWidth: 'sm',
    outlineStyle: 'solid',
    ':enabled:hover': {
      ...hoverCss,
    },
    ':not(active):focus': {
      ...focusCss,
    },
    ':enabled:active': {
      ...activeCss,
    },
    ':disabled': {
      color: 'b-400',
      backgroundColor: !isLink && 'muted',
      cursor: 'not-allowed',
    },
  };

  if (active) {
    return { ...baseCss, ...activeCss };
  }

  return baseCss;
};

export interface ButtonAddonCssParams {
  hasChildren: boolean;
  variant: ButtonVariant;
  size: Size;
}

export const buttonAddonCss = ({
  hasChildren,
  variant,
  size,
}: ButtonAddonCssParams): ThemeUIStyleObject => ({
  marginRight: hasChildren && 'sm',
  marginY: getPadding({ variant, size }),
});
