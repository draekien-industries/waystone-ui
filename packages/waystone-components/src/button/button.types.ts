export type ButtonVariant =
  | 'primary'
  | 'secondary'
  | 'accent'
  | 'highlight'
  | 'ghost'
  | 'link';
