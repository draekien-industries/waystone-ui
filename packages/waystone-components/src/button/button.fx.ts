import { addDarkModeSuffix, Size } from '@waystone/core';
import { ButtonVariant } from './button.styles';

export const getBackgroundColor = (
  variant: ButtonVariant,
  darkMode: boolean
) => {
  switch (variant) {
    case 'ghost':
      return darkMode ? 'b-700' : 'b-100';
    case 'link':
      return 'transparent';
    case 'primary':
    case 'secondary':
    case 'accent':
    case 'highlight':
      return `${variant[0]}-200`;
  }
};

export const getColor = (variant: ButtonVariant) => {
  switch (variant) {
    case 'primary':
    case 'secondary':
    case 'accent':
    case 'highlight':
      return 'b-600';
    default:
      return 'text';
  }
};

type ComputedBoxShadowStyles = {
  normal: string;
  hover: string;
  active: string;
};

export const getBoxShadow = (
  variant: ButtonVariant,
  darkMode: boolean
): ComputedBoxShadowStyles => {
  if (['ghost', 'link'].includes(variant)) {
    return {
      normal: 'none',
      hover: 'none',
      active: 'none',
    };
  }

  return {
    normal: addDarkModeSuffix('sm', darkMode),
    hover: addDarkModeSuffix('md', darkMode),
    active: addDarkModeSuffix('sm', darkMode),
  };
};

type ComputedPaddingStyles = {
  paddingX: string | number;
  paddingY: string | number;
};

export const getPadding = (
  size: Size,
  noPadding: boolean
): ComputedPaddingStyles => {
  if (noPadding) return { paddingX: 0, paddingY: 0 };

  switch (size) {
    case 'sm': {
      return {
        paddingX: 'md',
        paddingY: 'xs',
      };
    }
    case 'lg': {
      return {
        paddingX: 'xl',
        paddingY: '0.75rem',
      };
    }
    default:
      return {
        paddingX: 'lg',
        paddingY: 'sm',
      };
  }
};

export const getVariant = (size: Size) => {
  switch (size) {
    case 'sm':
      return 'text.caption';
    case 'lg':
      return 'text.eyebrow';
    default:
      return 'text.body';
  }
};
