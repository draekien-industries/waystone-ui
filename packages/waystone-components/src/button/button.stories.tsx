import React from 'react';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import { Button, ButtonProps } from './button';

export default {
  title: 'Components/Button',
  component: Button,
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />;

export const Primary = Template.bind({});

Primary.args = {
  children: 'Button',
} as ButtonProps;

export const WithIcon = Template.bind({});

WithIcon.args = {
  children: 'Button',
  icon: { variant: 'filled', name: 'check' },
} as ButtonProps;
