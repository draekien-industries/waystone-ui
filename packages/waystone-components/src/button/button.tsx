'use client';

import { ButtonVariant, Size } from '@waystone/core';
import { useMemo } from 'react';
import {
  Flex,
  Button as ThemeUiButton,
  ButtonProps as ThemeUiButtonProps,
} from 'theme-ui';
import {
  CanActivate,
  CanBeFullWidth,
  CanDisable,
  CanLoad,
  HasSize,
  HasVariant,
  HasWidth,
} from '../common';
import { useIsDarkMode } from '../hooks/useIsDarkMode';
import { Icon, IconProps } from '../icon';
import { Spinner } from '../spinner';
import { Text } from '../text';
import { getTextVariant } from './button.fx';
import {
  ButtonAddonCssParams,
  ButtonCssParams,
  buttonAddonCss,
  buttonCss,
} from './button.styles';

export interface ButtonProps
  extends Omit<ThemeUiButtonProps, 'sx' | 'variant' | 'children'>,
    HasVariant<ButtonVariant>,
    HasSize,
    HasWidth,
    CanDisable,
    CanLoad,
    CanActivate,
    CanBeFullWidth {
  /** The text to render inside a button. */
  children?: string;
  /** The icon and it's variant to render. */
  icon?: Pick<IconProps, 'variant' | 'name'>;
  /**
   * Should the button hide it's contents while it is in the loading state.
   * @default false
   */
  hideContentWhileLoading?: boolean;
}

/**
 * A pre-styled button with multiple sizes and variants. Supports loading, disabled, and active states.
 */
export const Button = ({
  children,
  icon,
  width,
  minWidth,
  maxWidth,
  fullWidth,
  variant = 'primary',
  size = 'sm',
  hideContentWhileLoading = false,
  loading = false,
  disabled = false,
  active = false,
  ...rest
}: ButtonProps) => {
  const isDarkMode = useIsDarkMode();

  const shouldBeDisabled = disabled || loading;
  const shouldRenderChildren =
    !!children && (!loading || !hideContentWhileLoading);
  const shouldRenderIcon = !loading && icon;

  const buttonSx = useMemo(() => {
    const buttonCssParams: ButtonCssParams = {
      variant,
      size,
      darkMode: isDarkMode,
      active,
      width,
      minWidth,
      maxWidth,
      fullWidth,
    };

    return buttonCss(buttonCssParams);
  }, [variant, size, isDarkMode, active, width, minWidth, maxWidth, fullWidth]);

  const buttonAddonSx = useMemo(() => {
    const buttonAddonCssParams: ButtonAddonCssParams = {
      variant,
      size,
      hasChildren: shouldRenderChildren,
    };

    return buttonAddonCss(buttonAddonCssParams);
  }, [variant, size, shouldRenderChildren]);

  const spinnerSize: Size = size === 'sm' ? 'sm' : 'md';

  return (
    <ThemeUiButton sx={buttonSx} disabled={shouldBeDisabled} {...rest}>
      {loading && (
        <Flex sx={buttonAddonSx}>
          <Spinner size={spinnerSize} />
        </Flex>
      )}
      {shouldRenderIcon && (
        <Flex sx={buttonAddonSx}>
          <Icon size="sm" {...icon} />
        </Flex>
      )}
      {shouldRenderChildren && (
        <Text variant={getTextVariant(size)} inline>
          {children}
        </Text>
      )}
    </ThemeUiButton>
  );
};
