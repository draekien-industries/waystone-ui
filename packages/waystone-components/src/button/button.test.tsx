import { render, screen } from '../__fixtures__';
import { Button } from './button';

describe('The Button component', () => {
  it('matches snapshot', () => {
    const { container } = render(<Button>button</Button>);

    expect(container).toMatchSnapshot();
  });

  it('renders a spinner when loading', () => {
    render(<Button loading>button</Button>);

    const spinner = screen.getByRole('progressbar');

    expect(spinner).toBeInTheDocument();
  });

  it('renders an icon when iconProps are provided', () => {
    render(<Button icon={{ name: 'access_alarm' }}>button</Button>);

    const icon = screen.getByText('access_alarm');

    expect(icon).toBeInTheDocument();
  });

  it('should be disabled when the disabled prop is true', () => {
    render(<Button disabled>Button</Button>);

    const button = screen.getByRole('button');

    expect(button).toBeDisabled();
  });

  it('should be disabled when the loading prop is true', () => {
    render(<Button loading>Button</Button>);

    const button = screen.getByRole('button');

    expect(button).toBeDisabled();
  });

  it('should use the provided widths', () => {
    render(
      <Button width="50%" minWidth="2rem" maxWidth="4rem">
        Button
      </Button>
    );

    const button = screen.getByRole('button');

    expect(button).toHaveStyle({
      width: '50%',
      'min-width': '2rem',
      'max-width': '4rem',
    });
  });

  it('should have 100% width when fullWidth prop is true', () => {
    render(<Button fullWidth>Button</Button>);

    const button = screen.getByRole('button');

    expect(button).toHaveStyle({
      width: '100%',
    });
  });

  it('renders a medium spinner when size is `md`', () => {
    render(
      <Button size="md" loading>
        button
      </Button>
    );

    const spinner = screen.getByRole('progressbar');

    expect(spinner).toHaveStyle({ height: '1.5em' });
  });
});
