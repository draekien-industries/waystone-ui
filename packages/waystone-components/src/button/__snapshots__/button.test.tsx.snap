// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Button component matches snapshot 1`] = `
<div>
  <button
    class="css-1ixzfme-Button"
  >
    <span
      class="css-3irkks-Text"
    >
      button
    </span>
  </button>
</div>
`;
